### DISCUSSION TITLE
Ruuvi Gateway

### DISCUSSION THREAD

**Discussion by ssahlender** [theengs/decoder#304] [Ideas]
Hello Community,

thanks for all the cool theengs tools, they are really great work.

I'm using right now OpenMQTTGateway and TheengsGateway and I'm I little bit unsure where to post my question.

Well let's start: As mentioned I play around with the OpenMQTTGateway and TheengsGateway, both are great when the sensor are locate nearby. But I have some sensors which are 30 meters away and the aren't seen with either of them.

I also own a [Ruuvi Gateway](https://ruuvi.com/gateway/#product-1) which has kind of weird looking antenna. But that one receives the signals from the sensor 30 meters away.
![image](https://user-images.githubusercontent.com/24881555/224363680-f3768ad4-60f3-45e4-bbfe-30aa8766b78f.png)
 
Also the software allows it to publish all ble messages the gateway receives in mqtt.

Here my idea / question starts:

How to decode it the best way?
- Take the decoder library and write some python around it, to read (raw) from and publish (decoded) message to mqtt
- Add a parameter to theengs gateway or openmqttgetway to connect to specific mqtt topic and decode / republish the messages
- Any other ideas? ( Just by that specific esp32 model, because it has a great antenne )
- OpenMQTTGateway firmware for Ruuvi Gateway

If need, I can post how the published topics from ruuvi gateway look like.

Thanks + Regards Stefan

**Comment by DigiH** [comment]
Hi @ssahlender,

interesting condition with your set-up. As Theengs Gateway is already able to receive undecoded MQTT messages from one or more OpenMQTTGateways, it would be possible to add the topic the Ruuvi Gateway publishes it's MQTT messages on to then have Theengs Gateway decode them.

 If the undecoded messages are similar to the direct Ruuvi Tag messages it might already work with the existing Ruuvi decoders, if not possible some new decoders would need to be implemented.

A good start would be if you could get your sensors close to OpenMQTTGateway, turn on [Publishing advertisement and advanced data](https://docs.openmqttgateway.com/use/ble.html#advanced-publishing-advertisement-and-advanced-data-default-false) and then post some messages directly from OMG plus MQTT messages for the same sensor from the Ruuvi Gateway for comparison.

> Any other ideas? ( Just by that specific esp32 model, because it has a great antenne )

Personally I like to use **ESP32-WROOM-32D** as they have a sticking out built in antenna or even the **ESP32-WROOM-32U** which allows for an external antenna with different gain to be attached. For my set-up these two models improve the reception over other ESP32 models, but I couldn't say if they would also cover your 30 meter distance.

> OpenMQTTGateway firmware for Ruuvi Gateway

Reading the specs of the Ruuvi Gateway, it is based on an ESP32, possibly connected and accessible via the USB-C port. But the feasibility of installing OpenMQTTGateway on it would need further investigation :)

**Reply by ssahlender** [reply]
Hi @DigiH 

so I tried what you mentioned to compare the messages:

{"id":"F1:52:A4:8A:61:D0","mac_type":1,"adv_type":0,"manufacturerdata":"9904050ea553ceb4a20020ffd003ec91160514b5f152a48a61d0","rssi":-80,"brand":"Ruuvi","model":"RuuviTag","model_id":"RuuviTag_RAWv2","type":"ACEL","tempc":18.745,"tempf":65.741,"hum":53.635,"pres":962.42,"accx":0.031381,"accy":-0.047072,"accz":0.984588,"volt":2.76,"tx":4,"mov":5,"seq":5301}

{
	"gw_mac":	"XX:XX:XX:XX:X:XX",
	"rssi":	-63,
	"aoa":	[],
	"gwts":	"1678469306",
	"ts":	"1678469306",
	"data":	"0201061BFF9904050EA653F3B4A20028FFD403EC917605152EF152A48A61D0",
	"coords":	""
}

So at the end the messages for one of my ruuvi tags are the same ;-) ... it's the MAC Address.
Above output is openmqttgateway below the output of what the ruuvi gateway sends ... topic is the mac address of the tag.

Interesting point is ... the node red node: https://github.com/ojousima/node-red can decode it directly from mqtt but it should also work with an usb stick connected to node red node ... so still wondering why the raw messages are so different.

Thanks for hint with these two ESP32 Models

Regards Stefan

**Reply by DigiH** [reply]
Hi @ssahlender,

I don't think they are actually that different, just with slightly different property values possibly due to minimal time differences of the reading/publishing

So from the Ruuvi gateway you get
`
"data": "0201061bff9904050ea653f3b4a20028ffd403ec917605152ef152a48a61d0",`

from OpenMQTTGateway

`"9904050ea553ceb4a20020ffd003ec91160514b5f152a48a61d0"`

from the Ruuvi gateway take away the prefix

0201061bff  **9904050ea653f3b4a20028ffd403ec917605152ef152a48a61d0** and you are left with something starting very similar to the manufacturerdata OMG decodes. 

With the temperature and humidity as an example with the following very slight differences

**Temperature**
OMG: 0ea5 > 3749 / 200 = 18,745 C°
Ruuvi: 0ea6 > 3750 / 200 = 18,75 C°

**Humidity**
OMG: 53ce > 21454 / 400 = 53,635 %
Ruuvi: 53f3 > 21491 / 400 = 53,7275 %

Haven't done any of the other properties, but I am confident that part of this part ;) of the Ruuvi Gateway "data" key data is identical to the manufacturerdata directly received by OMG and decoded.

What **topic** does the Ruuvi gateway publish on?

**Reply by ssahlender** [reply]
Thanks for the quick reply.

The ruuvi gateway publishes on: ruuvi/mac so ruuvi/AA:BB:CC:DD:EE:FF

So I also checked the manual of theengs gateway ... thanks for that hint.

And started my container with:

docker run --rm \
    --network host \
    -e MQTT_HOST=192.168.0.1 \
    -e MQTT_PUB_TOPIC=TheengsGateway/BTtoMQTT \
    -e MQTT_SUB_TOPIC=ruuvi \
    -e PUBLISH_ALL=true \
    -e LOG_LEVEL=DEBUG \
    --name TheengsGateway \
    theengs/gateway

I guess with the info you wrote I maybe play a little bit around with node red to republish it in the format mentioned here:
https://gateway.theengs.io/use/use.html#mqtttomqtt-decoding as well 

Regards

Stefan

**Reply by DigiH** [reply]
> I guess with the info you wrote I maybe play a little bit around with node red to republish it in the format mentioned here:
> https://gateway.theengs.io/use/use.html#mqtttomqtt-decoding as well

That sounds like your best bet, node-red to cut off the Ruuvi Gateway "data" key data prefix, the re-publish it shortened and still encoded with the MAC (taken from the end of the data) as the id in the Theengs format. Then your existing Theengs Gateway will decoder it as if it was directly from the sensor itself :)

When you test and verify the above please let us know how it is all working. If there is demand from others for this _via Ruuvi Gateway_ sensor decoding we could include this routine directly in Theengs Gateway in the future to eliminate the need for any node-red or similar step in between.

**Reply by ssahlender** [reply]
Sure, will post is as soon is I got it running,

so output of node-red is now:

{"id":"CE:52:7C:AF:5C:6B","mac_type":1,"manufacturerdata":"9904050e5d4670b7980030ffd403e09d5618305fce527caf5c6b","rssi":-54}

which should be the format that TheengsGateway could read.

Publishing it even to the default topic:

home/ruuvi/BTtoMQTT/undecoded/CE527CAF5C6B

Started the python version with:

.local/bin/TheengsGateway -H 192.168.0.1 -ll DEBUG

But nothing is in CLI Output / MQTT Output ... I guess some something I miss while reading the manual.

**Reply by DigiH** [reply]
How about with just the topic

`home/ruuvi/BTtoMQTT/undecoded`

?

**Reply by ssahlender** [reply]
Yup that was the fault ;-) ... After that TheengsGateway gets active and decoding works. So I'll cleanup my code a little bit and try it with my Xiaomi Sensors as well.

**Reply by DigiH** [reply]
Good to hear, so your Ruuvi gateway is tied into the local Theengs Gateway with Theengs Decoder decoding :) 

Are the Xiaomi sensors being picked up by the Ruuvi Gateway over a longer distance as well?

**Reply by ssahlender** [reply]
@DigiH 
Yes, even the xiaomi are picked up over longer distance.

But also some confusion about decoding:
output of openmqttgateway
{"id":"58:2D:34:3B:8B:17","mac_type":0,"adv_type":0,"rssi":-91,"servicedata":"5020aa01f4178b3b342d58061002d201","servicedatauuid":"0xfe95","brand":"Xiaomi","model":"Mi Jia round","model_id":"LYWSDCGQ","type":"THB","hum":46.6}

Somehow theengs gateway doesn't decode that one.
{"id":"58:2D:34:3B:8B:17","mac_type":1,"manufacturerdata":"95fe5020aa01a8178b3b342d58061002d201","ruuvi_raw":"95FE5020AA01A8178B3B342D58061002D201","rssi":-73}

Looks like for every manufacture ruuvi gateway adds a prefix:

Ruvvi Tags:
0201061BFF

Xiaomi Mia Round:
0201061316

**Reply by DigiH** [reply]
@ssahlender 

> Somehow theengs gateway doesn't decode that one. {"id":"58:2D:34:3B:8B:17","mac_type":1,"manufacturerdata":"95fe5020aa01a8178b3b342d58061002d201","ruuvi_raw":"95FE5020AA01A8178B3B342D58061002D201","rssi":-73}

Isn't this one not split up fully and has the wrong "manufacturerdata" key? 
The _manufacturerdata_ (should really be servicedata(plus uuid) here from the Ruuvi "data" should be split up into the leading 95fe as little endian serviceuuid hex prefixed, i.e. "0xfe95", then servicedata as the remaining "5020aa01a8178b3b342d58061002d201". As shown above the for the correctly decoded OMG result.

Then it should be correctly recognised and decoded.

So some devices' Ruuvi data needs to be node-red converted into manufacturerdata only, others as servicedatauuid plus servicedata.

Best to look at the relevant Decoder files for what is expected for different devices, or as you did with the direct OMG decoding.

https://github.com/theengs/decoder/tree/development/src/devices

> Looks like for every manufacture ruuvi gateway adds a prefix:
> 
> Ruvvi Tags: 0201061BFF
> 
> Xiaomi Mia Round: 0201061316

Good find. I wonder if this is generally the case for different snesors or possibly varying with different Ruuvi Gateways as internal IDs, or even with the same Ruuvi Gateway after a restart of the gateway.

**Reply by ssahlender** [reply]
@DigiH Just googled the 0201061BFF: https://f.ruuvi.com/t/parsing-ruuvi-rawdata-in-java/6026 ... looks like that it's also parseable as described in that post. For the other things I'll try to extend my node-red code.

**Reply by ssahlender** [reply]
So, looks like theengs gateway even decodes this:
{"id":"58:2D:34:3B:8B:17","mac_type":1,"servicedata":"5020aa0194178b3b342d58061002f901","servicedatauuid":"95FE","ruuvi_raw":"020106131695FE5020AA0194178B3B342D58061002F901","rssi":-73}

95FE insted of 0xfe95

My code so far, before that I just filter out just the ruuvi and xiaomi messages
```
var mac_short = msg.topic.replaceAll(":","")
var sduuid = ""
var ble_manufacturer = msg.payload.data.slice(0,10)

var payload_json = ""
var data_short = ""

// ruuvi
if (ble_manufacturer == "0201061BFF")
{
    data_short = msg.payload.data.slice(10)
    payload_json = {
        "id": msg.topic,
        "mac_type": 1 ,
        "manufacturerdata": data_short.toLowerCase(),
        "ruuvi_raw": msg.payload.data,
        "rssi": msg.payload.rssi,
    }
}

// xiaomi
if (ble_manufacturer == "0201061316")
{
    data_short = msg.payload.data.slice(14)
    sduuid = msg.payload.data.slice(10,14)
    payload_json = {
        "id": msg.topic,
        "mac_type": 1,
        "servicedata": data_short.toLowerCase(),
        "servicedatauuid": sduuid,
        "ruuvi_raw": msg.payload.data,
        "rssi": msg.payload.rssi,
    }
}


var m_array = [
    {
        payload: payload_json,
        topic: "ruuvi2theengs",
    },
 //   {
 //       payload: payload_json,
 //       topic: "ruuvi2theengs/"+mac_short,
 //   },
];
return [m_array];
```

**Reply by koenvervloesem** [reply]
These are BLE advertising data structures in Length/Type/Value (LTV) format:

02-01-06: Length 2, Type 1 (Flags), Data: 06 (with length 2 - 1 for the Type byte)
1B-FF-DATA: Length 27, Type FF (Manufacturer Specific Data), Data (with length 26)
13-16-DATA: Length 19, Type 16 (Service Data - 16-bit UUID), Data (with length 18)

So if you want to correctly decode the value of the `data`  key from your Ruuvi Gateway's output, you need to cut this data into its advertising data structures and then convert the structure for type FF to a `manufacturerdata` key to send to Theengs Gateway, and convert the structure for type 16 to `servicedatauuid` and `servicedata` keys. See section 2.3 of the [Assigned Numbers Document](https://www.bluetooth.com/specifications/assigned-numbers/) for the full list.

We can probably add support for this in Theengs Gateway, but maybe try this first in Node-RED and see whether this approach works.

Edit: You already figured this out partly while I was writing this :-)

**Reply by ssahlender** [reply]
@koenvervloesem Thanks for feedback, well let's say i kind of did the hardcoded implementation ;-). Thanks for your feedback.

So in case to skip the complete node-red stuff in between, ruuvi gateway output looks like this before it's processed by node-red:

For xiaomi from MQTT Explorer:
![image](https://user-images.githubusercontent.com/24881555/224550187-0fa9f3fa-b8a5-4bd2-a2dd-686f3b9ed7d3.png)

For ruuvitags from MQTT Explorer:
![image](https://user-images.githubusercontent.com/24881555/224550262-92e79bd4-db9e-42f4-9256-23adea0b3afa.png)

As also said to @DigiH  the ruuvi gateway antenna is really great. I live in a house from the 1930s with very thick walls but the gateway catches it all. The xiaomi sensors are distributed over 4 floors, as the ruuvi tags are. Two of the are outside a little bit away from the building.

Just don't know how many of the Theengs / OpenMQTTGateway users own one of this gateways since it's kind of expensive when you didn't get it during the kickstartet campaign. It's around 200 Euro.

Regards Stefan

**Reply by DigiH** [reply]
@ssahlender Thanks for all the info and trying out the node-red conversion implementation which looks like it's working really nicely for you!

While it's true that possibly not many other users own the Ruuvi Gateway, as Theengs Gateway already implements decoding undecoded messages from OpenMQTTGateways, it could well be worth looking into adding the additional Ruuvi Gateway decoding to Theengs Gateway, if @koenvervloesem and the others in the team agree.

Are you planning to publish your node-red implementation for others to use the same?

**Reply by ssahlender** [reply]
@DigiH 

Well I could put my flow on github and publish it to the node red website ... also with the info, that it's WIP and the algorithm has lot of room for improvement.
Will write here when it's finished. Just give me some days  have to learn markdown a little bit ;-).

Regards

Stefan

**Reply by DigiH** [reply]
Hi @ssahlender 

> So, looks like theengs gateway even decodes this: {"id":"58:2D:34:3B:8B:17","mac_type":1,"servicedata":"5020aa0194178b3b342d58061002f901","servicedatauuid":"95FE","ruuvi_raw":"020106131695FE5020AA0194178B3B342D58061002F901","rssi":-73}
> 
> 95FE insted of 0xfe95

This is being decoded because this particular LYWSDCGQ decoder does not actually use the serviceuuid in it's model condition. If it would have, or any other decoders using the actual serviceuuid, of which there are many, it will fail if not in reversed little endian order. The 0x prefixing should not matter that much AFAIK, without having looked at it more closely right now.

Just FYI for a correct serviceuuid recognition of your node-red implementation.
